### PR TITLE
BLD: Build wheels with cibuildwheel 2.12.0

### PR DESCRIPTION
**Pull request by charris**
The 2.11.4 version of cibuildwheel was failing on 32 bit windows. This is to check that and provide an error log that can be linked to in a cibuildwheel issue.

### CODE REVIEW

**Comment by henryiii** [comment]
What about adding a `os.makedirs(os.path.join(sys.prefix, 'libs'), exist_ok=True)` right above https://github.com/numpy/numpy/blob/46142454ed4746ac991d4074d2943d892ede35d6/numpy/distutils/mingw32ccompiler.py#L441?

**Comment by charris** [comment]
> What about adding a os.makedirs(os.path.join(sys.prefix, 'libs'), exist_ok=True)

I'm a bit hesitant without knowing _why_ the current code breaks in 2.11.3.  I wonder if the meson build also needs to deal with this. @rgommers Thoughts?

**Comment by henryiii** [comment]
Running in bash was likely a more GNU-like environment than cmd. Meson behaves very differently if you switch between bash & cmd - this is primarily why we changed this default to cmd in cibuildwheel. In bash it was not able to find MSBC, and was stuck with mingw.

You can manually test the last two patch versions with bash by running it via pipx instead of using the action (in a bash run action).

**Comment by rgommers** [comment]
So the issue is:
```
FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\runneradmin\\AppData\\Local\\Temp\\cibw-run-oedjqo95\\cp39-win32\\build\\venv\\libs\\python39.def'
```

> I'm a bit hesitant without knowing _why_ the current code breaks in 2.11.3. I wonder if the meson build also needs to deal with this. @rgommers Thoughts?

@henryiii's suggestion seems fine to me - whatever works here I'd say. Meson build: probably not affected by this, since it's choking on something in `numpy.distutils` and we won't be using that anymore in the near future.

**Comment by charris** [comment]
@henryiii That didn't fix things, but pushed the failure further on to link time. Looks like the Python library wasn't available.

**Comment by henryiii** [comment]
Somehow it's not resolving in the non-bash environment correctly. As @joerick points out in the cibuildwheel issue, in https://github.com/numpy/numpy/issues/20391, I think this is known to be fragile.

It looks like it goes through a bit of work to detect the library path: https://github.com/numpy/numpy/blob/46142454ed4746ac991d4074d2943d892ede35d6/numpy/distutils/mingw32ccompiler.py#L363-L388, but then it ignores all that and just puts the .def in a hardcoded sys.prefix / "libs" dir. Since that doesn't exist, I'm assuming that it found the library in a different dir.

**Comment by charris** [comment]
close/reopen to rerun.

**Comment by charris** [comment]
The macosx failures are new and probably due to a rebase that exposed recent numpy SIMD merges. I suspect this show up here due to a  change in compiler versions. @henryiii Has the macosx compiler been updated in cibuildwheel? @seiko2plus Thoughts?

**Comment by mattip** [comment]
The clang compiler [does not recognize](https://github.com/numpy/numpy/actions/runs/3868384006/jobs/6593786865#step:5:6729) the `asm()` calls from #22954. Note that the CI for that PR ran on azure, this is running on github actions. The [image software](https://github.com/numpy/numpy/actions/runs/3868384006/jobs/6593786865#step:1:10) in this PR is different from [the other PR](https://dev.azure.com/numpy/numpy/_build/results?buildId=27548&view=logs&j=04e3746b-8857-5383-a5f1-d85228b91063&t=7b62d4a1-512d-4db0-a24e-38c958e2a0bd&l=9): this one is using clang 14.0.0 where the other is using [clang 13.0.0](https://dev.azure.com/numpy/numpy/_build/results?buildId=27548&view=logs&jobId=f758ecd3-9bca-5197-a703-541c3e314cdb&j=04e3746b-8857-5383-a5f1-d85228b91063&t=148e1615-91c7-5971-6068-868e18041738).

Also for some reason the 32-bit windows builds are [linking with g++](https://github.com/numpy/numpy/actions/runs/3868384006/jobs/6593787275#step:5:1450) instead of MSVC.

**Comment by mattip** [comment]
Thanks @seiko2plus, now all that is left is the win32 failures. [Here](https://github.com/numpy/numpy/actions/runs/3869161932) was the last successful run. I wonder what changed.

**Comment by mattip** [comment]
The real differences start just after `running build_clib`. The [bad run](https://github.com/numpy/numpy/actions/runs/3871798937/jobs/6599931225#step:5:1040) prefers the mingw 32-bit compiler, while the [good run](https://github.com/numpy/numpy/actions/runs/3869161932/jobs/6595092303#step:5:1007) prefers the MSVC compiler.

**Comment by mattip** [comment]
The 64-bit build uses MSVC in both cases, [here](https://github.com/numpy/numpy/actions/runs/3871798937/jobs/6599930983#step:5:993) is the `running build_clib` part of the log for the successful 64-bit build

**Comment by seiko2plus** [comment]
I tried to install python-32bit but still got 64-bit version :
```Bash
Run pipx run --python "C:\hostedtoolcache\windows\Python\3.11.1\x64\python.exe" --spec "D:\a\_actions\pypa\cibuildwheel\v2.11.3" cibuildwheel "." --output-dir '"wheelhouse"' --config-file '""' --only '""'
```

**Comment by seiko2plus** [comment]
the linker wasn't able to identify 32bit of libpython
```Bash
2023-01-09T11:30:39.9452849Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x6f): undefined reference to `_imp___Py_NoneStruct'
2023-01-09T11:30:39.9463148Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x252): undefined reference to `_imp___Py_NoneStruct'
2023-01-09T11:30:39.9464256Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x2de): undefined reference to `_imp__PyType_IsSubtype'
2023-01-09T11:30:39.9465285Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x2ff): undefined reference to `_imp___Py_NoneStruct'
2023-01-09T11:30:39.9466173Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x369): undefined reference to `_imp__PyExc_TypeError'
2023-01-09T11:30:39.9467149Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x374): undefined reference to `_imp__PyErr_SetString'
2023-01-09T11:30:39.9468031Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x3b5): undefined reference to `_imp___PyArg_ParseTuple_SizeT'
2023-01-09T11:30:39.9468927Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x3c3): undefined reference to `_imp__PyFloat_Type'
2023-01-09T11:30:39.9471250Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x3d4): undefined reference to `_imp__PyType_IsSubtype'
2023-01-09T11:30:39.9481568Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x3e7): undefined reference to `_imp__PyComplex_Type'
2023-01-09T11:30:39.9482597Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x408): undefined reference to `_imp__PyBool_Type'
2023-01-09T11:30:39.9483479Z   C:/rtools40/mingw32/bin/../lib/gcc/i686-w64-mingw32/8.3.0/../../../../i686-w64-mingw32/bin/ld.exe: build\temp.win32-3.9\Release\build\src.win32-3.9\numpy\core\src\multiarray\_multiarray_tests.o:_multiarray_tests.c:(.text+0x421):
```

**Comment by mattip** [comment]
Did you convince the build process to use MSVC and not g++?

**Comment by seiko2plus** [comment]
>Did you convince the build process to use MSVC and not g++?

passed after uses MSVC

File: numpy/distutils/mingw32ccompiler.py [hunk 1]
```diff
@@ -440,6 +440,7 @@ def _build_import_library_x86():
     from numpy.distutils import lib2def
 
     def_name = "python%d%d.def" % tuple(sys.version_info[:2])
+    os.makedirs(os.path.join(sys.prefix, 'libs'), exist_ok=True)
```
Comment (henryiii) [inline numpy/distutils/mingw32ccompiler.py:443]: Is this still needed now? I'd assume not?

File: tools/ci/cirrus_general.yml [hunk 2]
```diff
@@ -1,6 +1,6 @@
 build_and_store_wheels: &BUILD_AND_STORE_WHEELS
   install_cibuildwheel_script:
-    - python -m pip install cibuildwheel==2.11.2
+    - python -m pip install cibuildwheel==2.11.3
```
Comment (henryiii) [inline tools/ci/cirrus_general.yml:3]: Can go back to 2.11.4

Comment (charris) [reply tools/ci/cirrus_general.yml]: Done.

File: numpy/distutils/mingw32ccompiler.py (hunk 1 above)
Comment (charris) [reply numpy/distutils/mingw32ccompiler.py]: No.

**Comment by charris** [comment]
@henryiii Updating to 2.11.4 brings back the pp38 segfault on manylinux2014.

**Comment by mattip** [comment]
That was fixed in [PyPy issue 3870](https://foss.heptapod.net/pypy/pypy/-/issues/3870) which was part of the PyPy v7.3.11 release (cibuildwheel 2.11.4 uses v7.3.10). There is a PR to upgrade PyPy in pypa/cibuildwheel#1372

**Comment by charris** [comment]
@mattip Thanks for the info. Looks like the next step is to wait for cibuildwheel 2.11.5 and upgrade the version again. Or we could put this in with 2.11.3, but there is no hurry.

**Comment by charris** [comment]
Updated to use cibuildwheel 2.12.0 released four hours ago.

**Comment by charris** [comment]
Self merging so work on top of this can proceed. @mattip You should be able to use pp39 now.

**Pull request merged**